File: src/sdk/rust/tests/subscribe_test.rs
// subscribe_test.rs
//
// Unit tests for event subscription message types.
// These tests verify serialization/deserialization and event model correctness
// without requiring a running server.

#[cfg(test)]
mod tests {
    use serde::{Deserialize, Serialize};
    use serde_json::json;
    use std::collections::HashMap;

    const EVENT_URI: &str = "/appmesh/event";

    #[derive(Debug, Serialize, Deserialize, PartialEq)]
    struct AppEvent {
        subscription_id: String,
        event_type: String,
        app_name: String,
        timestamp: i64,
        sequence: u64,
        data: serde_json::Value,
    }

    #[test]
    fn test_app_event_deserialization() {
        let raw = r#"{
            "subscription_id": "sub123",
            "event_type": "EXIT",
            "app_name": "myapp",
            "timestamp": 1714000000,
            "sequence": 42,
            "data": {"exit_code": 1, "pid": 12345}
        }"#;

        let event: AppEvent = serde_json::from_str(raw).expect("Failed to deserialize AppEvent");
        assert_eq!(event.subscription_id, "sub123");
        assert_eq!(event.event_type, "EXIT");
        assert_eq!(event.app_name, "myapp");
        assert_eq!(event.timestamp, 1714000000);
        assert_eq!(event.sequence, 42);
        assert_eq!(event.data["exit_code"], 1);
        assert_eq!(event.data["pid"], 12345);
    }

    #[test]
    fn test_app_event_serialization_roundtrip() {
        let event = AppEvent {
            subscription_id: "sub-rt".to_string(),
            event_type: "START".to_string(),
            app_name: "test-app".to_string(),
            timestamp: 1714000000,
            sequence: 1,
            data: json!({"pid": 9999, "process_uuid": "abc-def"}),
        };

        let serialized = serde_json::to_string(&event).expect("Failed to serialize");
        let deserialized: AppEvent = serde_json::from_str(&serialized).expect("Failed to deserialize");
        assert_eq!(event, deserialized);
    }

    #[test]
    fn test_event_uri_identification() {
        assert_eq!(EVENT_URI, "/appmesh/event");

        // An event response is identified by request_uri == EVENT_URI
        let response_uri = "/appmesh/event";
        assert!(response_uri == EVENT_URI, "Event responses use the EVENT_URI sentinel");

        let normal_uri = "/appmesh/app/test";
        assert!(normal_uri != EVENT_URI, "Normal responses should not match EVENT_URI");
    }

    #[test]
    fn test_event_types() {
        let valid_types = vec![
            "START",
            "EXIT",
            "STDOUT",
            "HEALTH",
            "STATUS",
            "REMOVED",
        ];

        for event_type in &valid_types {
            let event = AppEvent {
                subscription_id: "sub-1".to_string(),
                event_type: event_type.to_string(),
                app_name: "test".to_string(),
                timestamp: 0,
                sequence: 0,
                data: json!({}),
            };
            let json_str = serde_json::to_string(&event).expect("serialize");
            assert!(json_str.contains(event_type));
        }
    }

    #[test]
    fn test_event_with_headers() {
        let mut headers: HashMap<String, String> = HashMap::new();
        headers.insert("X-Subscription-Id".to_string(), "sub-hdr".to_string());
        headers.insert("X-Event-Type".to_string(), "EXIT".to_string());
        headers.insert("X-App-Name".to_string(), "myapp".to_string());

        assert_eq!(headers.get("X-Subscription-Id"), Some(&"sub-hdr".to_string()));
        assert_eq!(headers.get("X-Event-Type"), Some(&"EXIT".to_string()));
        assert_eq!(headers.get("X-App-Name"), Some(&"myapp".to_string()));
    }

    #[test]
    fn test_subscribe_result_deserialization() {
        let raw = r#"{
            "subscription_id": "cqk8g7l4d",
            "app_name": "myapp",
            "events": ["START", "EXIT", "STDOUT"]
        }"#;

        #[derive(Debug, Deserialize)]
        struct SubscriptionResult {
            subscription_id: String,
            app_name: String,
            events: Vec<String>,
        }

        let result: SubscriptionResult = serde_json::from_str(raw).expect("Failed to deserialize");
        assert_eq!(result.subscription_id, "cqk8g7l4d");
        assert_eq!(result.app_name, "myapp");
        assert_eq!(result.events.len(), 3);
        assert!(result.events.contains(&"STDOUT".to_string()));
    }
}

File: src/sdk/java/src/test/java/appmesh/SubscribeTest.java
package appmesh;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;

/**
 * Unit tests for event subscription message types.
 * These tests verify serialization/deserialization and event model correctness
 * without requiring a running server.
 */
public class SubscribeTest {

    private static final String EVENT_URI = "/appmesh/event";

    @Test
    public void testResponseMessageDeserialization() throws IOException {
        // Build a ResponseMessage with event payload
        ResponseMessage original = new ResponseMessage();
        original.uuid = "evt-uuid-1";
        original.request_uri = EVENT_URI;
        original.http_status = 200;
        original.body_msg_type = "application/json";

        JSONObject eventBody = new JSONObject();
        eventBody.put("subscription_id", "sub-test");
        eventBody.put("event_type", "EXIT");
        eventBody.put("app_name", "myapp");
        eventBody.put("timestamp", 1714000000);
        eventBody.put("sequence", 42);
        eventBody.put("data", new JSONObject().put("exit_code", 1).put("pid", 12345));

        original.body = eventBody.toString().getBytes("UTF-8");
        original.headers = new HashMap<>();
        original.headers.put("X-Subscription-Id", "sub-test");
        original.headers.put("X-Event-Type", "EXIT");
        original.headers.put("X-App-Name", "myapp");

        // Serialize and deserialize
        byte[] serialized = original.serialize();
        assertNotNull(serialized);
        assertTrue(serialized.length > 0);

        ResponseMessage decoded = ResponseMessage.deserialize(serialized);
        assertEquals("evt-uuid-1", decoded.uuid);
        assertEquals(EVENT_URI, decoded.request_uri);
        assertEquals(200, decoded.http_status);
        assertEquals("sub-test", decoded.headers.get("X-Subscription-Id"));
        assertEquals("EXIT", decoded.headers.get("X-Event-Type"));
        assertEquals("myapp", decoded.headers.get("X-App-Name"));

        // Parse the body back as JSON
        String bodyStr = new String(decoded.body, "UTF-8");
        JSONObject parsedBody = new JSONObject(bodyStr);
        assertEquals("sub-test", parsedBody.getString("subscription_id"));
        assertEquals("EXIT", parsedBody.getString("event_type"));
        assertEquals("myapp", parsedBody.getString("app_name"));
        assertEquals(1, parsedBody.getJSONObject("data").getInt("exit_code"));
    }

    @Test
    public void testEventUriConstant() {
        assertEquals("/appmesh/event", EVENT_URI);
    }

    @Test
    public void testRequestMessageForSubscribe() throws IOException {
        // Build a subscribe request
        RequestMessage req = new RequestMessage();
        req.uuid = "req-uuid-sub";
        req.request_uri = "/appmesh/app/myapp/subscribe";
        req.http_method = "POST";
        req.client_addr = "test-client";
        req.headers = new HashMap<>();
        req.headers.put("Authorization", "Bearer test-token");
        req.query = new HashMap<>();
        req.query.put("events", "START,EXIT,STDOUT");

        byte[] serialized = req.serialize();
        assertNotNull(serialized);
        assertTrue(serialized.length > 0);
    }

    /** Scripted mock transport: framed messages fed through a queue; an empty array reads as EOF. */
    private static MessageDemuxer.MessageReader queueReader(BlockingQueue<byte[]> queue) {
        return () -> {
            byte[] data = queue.take();
            return data.length == 0 ? null : data;
        };
    }

    // Conformance: S7 (partial) — waiter registered before the response arrives is routed by UUID (docs/source/SDKContract.md)
    @Test
    public void testDemuxerRoutesResponseToPreRegisteredWaiter() throws Exception {
        BlockingQueue<byte[]> queue = new LinkedBlockingQueue<>();
        MessageDemuxer demuxer = new MessageDemuxer(queueReader(queue));
        demuxer.start();
        try {
            ResponseMessage resp = new ResponseMessage();
            resp.uuid = "req-s7";
            resp.request_uri = "/appmesh/app/test";
            resp.http_status = 200;
            resp.body = "{}".getBytes("UTF-8");
            resp.headers = new HashMap<>();

            demuxer.registerPending("req-s7");
            queue.put(resp.serialize()); // response arrives right after "send"

            ResponseMessage got = demuxer.waitForResponse("req-s7");
            assertNotNull(got, "response arriving right after send must not be dropped");
            assertEquals("req-s7", got.uuid);
            assertEquals(200, got.http_status);
        } finally {
            demuxer.stop();
        }
    }

    // Conformance: S2 (demuxer) — transport EOF broadcasts __disconnected__ to every callback
    // and wakes pending waiters with null instead of hanging (docs/source/SDKContract.md)
    @Test
    public void testDemuxerDisconnectBroadcastUnblocksWaiters() throws Exception {
        BlockingQueue<byte[]> queue = new LinkedBlockingQueue<>();
        MessageDemuxer demuxer = new MessageDemuxer(queueReader(queue));
        demuxer.start();

        final CountDownLatch disconnected = new CountDownLatch(1);
        final AtomicReference<MessageDemuxer.AppEvent> received = new AtomicReference<>();
        demuxer.registerEventCallback("sub-s2", event -> {
            received.set(event);
            disconnected.countDown();
        });

        demuxer.registerPending("req-s2");
        final CountDownLatch waiterDone = new CountDownLatch(1);
        final AtomicReference<ResponseMessage> waiterResult = new AtomicReference<>();
        Thread waiter = new Thread(() -> {
            waiterResult.set(demuxer.waitForResponse("req-s2"));
            waiterDone.countDown();
        });
        waiter.start();

        queue.put(new byte[0]); // transport EOF

        assertTrue(disconnected.await(5, TimeUnit.SECONDS), "disconnect broadcast not delivered");
        assertEquals(MessageDemuxer.EVENT_TYPE_DISCONNECTED, received.get().eventType);
        assertEquals("sub-s2", received.get().subscriptionId);

        assertTrue(waiterDone.await(5, TimeUnit.SECONDS), "pending waiter not woken on disconnect");
        assertNull(waiterResult.get(), "disconnect wakes the waiter with null, never a response");
    }

    @Test
    public void testEventResponseHeaders() throws IOException {
        ResponseMessage resp = new ResponseMessage();
        resp.uuid = "evt-2";
        resp.request_uri = EVENT_URI;
        resp.http_status = 200;
        resp.body = "{}".getBytes("UTF-8");
        resp.headers = new HashMap<>();
        resp.headers.put("X-Subscription-Id", "sub-123");
        resp.headers.put("X-Event-Type", "HEALTH");
        resp.headers.put("X-App-Name", "test-app");

        byte[] buf = resp.serialize();
        ResponseMessage decoded = ResponseMessage.deserialize(buf);

        assertEquals("sub-123", decoded.headers.get("X-Subscription-Id"));
        assertEquals("HEALTH", decoded.headers.get("X-Event-Type"));
        assertEquals("test-app", decoded.headers.get("X-App-Name"));
    }
}
